### PR TITLE
[mlir][[spirv] Add support for math.log2 and math.log10 to GLSL/OpenCL SPIRV Backends

### PR DESCRIPTION
**Pull request by catnipsy**
As log2 and log10 are not available in spirv, realize them as a decomposition using spirv.CL.log/spirv.GL.Log.

### CODE REVIEW

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 1]
```diff
@@ -291,6 +291,65 @@ struct Log1pOpPattern final : public OpConversionPattern<math::Log1pOp> {
...
+  using OpConversionPattern<MathLogOp>::OpConversionPattern;
+  using typename OpConversionPattern<MathLogOp>::OpAdaptor;
+
+  LogicalResult
+  matchAndRewrite(MathLogOp operation, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    assert(adaptor.getOperands().size() == 1);
+    if (LogicalResult res = checkSourceOpTypes(rewriter, operation);
+        failed(res))
+      return res;
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:317]: Could you make the error more descriptive and use `return operation.notifyMatchFailure() << "Something failed";`?

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 2]
```diff
@@ -291,6 +291,65 @@ struct Log1pOpPattern final : public OpConversionPattern<math::Log1pOp> {
...
+
+    Location loc = operation.getLoc();
+    Type type = this->getTypeConverter()->convertType(operation.getType());
+    if (!type)
+      return failure();
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:322]: Also here, notify that type conversion failed

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 3]
```diff
@@ -291,6 +291,65 @@ struct Log1pOpPattern final : public OpConversionPattern<math::Log1pOp> {
...
+
+    auto getConstantValue = [&](double value) {
+      if (auto floatType = llvm::dyn_cast<FloatType>(type)) {
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:325]:
```suggestion
      if (auto floatType = dyn_cast<FloatType>(type)) {
```

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 4]
```diff
@@ -291,6 +291,65 @@ struct Log1pOpPattern final : public OpConversionPattern<math::Log1pOp> {
...
+        return rewriter.create<spirv::ConstantOp>(
+            loc, type, rewriter.getFloatAttr(floatType, value));
+      }
+      if (auto vectorType = llvm::dyn_cast<VectorType>(type)) {
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:329]:
```suggestion
      if (auto vectorType = dyn_cast<VectorType>(type)) {
```

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 5]
```diff
@@ -291,6 +291,65 @@ struct Log1pOpPattern final : public OpConversionPattern<math::Log1pOp> {
...
+              loc, type,
+              DenseFPElementsAttr::get(
+                  vectorType, FloatAttr::get(elemType, value).getValue()));
+        }
+      }
+
+      llvm_unreachable("unimplemented types for log2/log10");
+    };
+
+    auto constantValue = getConstantValue(
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:343]:
```suggestion
    Value constantValue = getConstantValue(
```

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 6]
```diff
@@ -291,6 +291,65 @@ struct Log1pOpPattern final : public OpConversionPattern<math::Log1pOp> {
...
+        std::is_same<MathLogOp, math::Log2Op>() ? LOG2_RECIPROCAL
+                                                : LOG10_RECIPROCAL);
+    auto log = rewriter.create<SpirvLogOp>(loc, adaptor.getOperand());
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:346]:
```suggestion
    Value log = rewriter.create<SpirvLogOp>(loc, adaptor.getOperand());
```

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp [hunk 7]
```diff
@@ -291,6 +291,65 @@ struct Log1pOpPattern final : public OpConversionPattern<math::Log1pOp> {
...
+///
+/// SPIR-V does not have direct operations for log2 and log10. Explicitly
+/// lower to these operations using:
+///   log2(x) = log(x) * 1/log(2)
+///   log10(x) = log(x) * 1/log(10)
+
+#define LOG2_RECIPROCAL                                                        \
+  1.442695040888963407359924681001892137426645954152985934135449407
+#define LOG10_RECIPROCAL                                                       \
+  0.4342944819032518276511289189166050822943970058036665661144537832
```
Comment (kuhar) [inline mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:304]: You can define these at the class level with `static constexpr double xyz = 1.442...;`

**Review by kuhar — COMMENTED** [review]
Thanks, this looks very useful. Just a couple minor issues that I see.

**Comment by kuhar** [comment]
@angelz913 could you verify that the math check outs?

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp (hunk 1 above)
Comment (meehatpa) [reply mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp:317]: `checkSourceOpTypes` returns `notifyMatchFailure` for a failure case.

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp (hunk 2 above)
Comment (meehatpa) [reply mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp]: Done

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp (hunk 3 above)
Comment (meehatpa) [reply mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp]: Done

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp (hunk 4 above)
Comment (meehatpa) [reply mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp]: Done

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp (hunk 5 above)
Comment (meehatpa) [reply mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp]: Done

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp (hunk 6 above)
Comment (meehatpa) [reply mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp]: Done

File: mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp (hunk 7 above)
Comment (meehatpa) [reply mlir/lib/Conversion/MathToSPIRV/MathToSPIRV.cpp]: Done

**Comment by angelz913** [comment]
> @angelz913 could you verify that the math check outs?

Looks correct to me

**Review by kuhar — APPROVED** [review]
LGTM

**Comment by meehatpa** [comment]
Thanks for the reviews!
Please merge :)

**Pull request merged**